src/transaction.ts

```diff
@@ -22,26 +22,60 @@ import {google} from '../protos/protos';
 
 import {Datastore, TransactionOptions} from '.';
 import {entity, Entity, Entities} from './entity';
-import {Query} from './query';
+import {
+  Query,
+  RunQueryCallback,
+  RunQueryInfo,
+  RunQueryOptions,
+  RunQueryResponse,
+} from './query';
 import {
   CommitCallback,
   CommitResponse,
   DatastoreRequest,
   RequestOptions,
   PrepareEntityObjectResponse,
+  CreateReadStreamOptions,
+  GetResponse,
+  GetCallback,
+  RequestCallback,
 } from './request';
 import {AggregateQuery} from './aggregate';
+import {Mutex} from 'async-mutex';
+
+type RunQueryResponseOptional = [
+  Entity[] | undefined,
+  RunQueryInfo | undefined,
+];
 
-// RequestPromiseReturnType should line up with the types in RequestCallback
-interface RequestPromiseReturnType {
+/**
+ * This type matches data typically passed into a callback the user supplies.
+ * The data matches promises created from an argument of Resolver<T> type.
+ */
+interface UserCallbackData<T> {
   err?: Error | null;
-  resp: any; // TODO: Replace with google.datastore.v1.IBeginTransactionResponse and address downstream issues
+  resp?: T;
 }
-interface RequestResolveFunction {
-  (callbackData: RequestPromiseReturnType): void;
+
+/**
+ * This is a type that matches the argument for a promise's resolve function.
+ * It is also constrained to match data returned that may contain an error.
+ */
+interface PromiseResolveFunction<T> {
+  (value: UserCallbackData<T> | PromiseLike<UserCallbackData<T>>): void;
 }
-interface RequestAsPromiseCallback {
-  (resolve: RequestResolveFunction): void;
+
+/**
+ * This is a type that matches the argument passed in when building a promise.
+ * It is also assures that the promise will resolve with data of PassThroughReturnType<T> type.
+ */
+interface Resolver<T> {
+  (resolve: PromiseResolveFunction<T>): void;
+}
```

Comment:
These types are all private, right?

Reply:
Yup. To make them available outside this module you would have to add `export` before the word `interface`.